docs/runtime/ir-caching.md

```diff
@@ -0,0 +1,209 @@
+   can use a path within that for caching. This should be
+   `$package/.enso/cache/ir/enso-$version/`, and can be got at by extending the
+   `pkg` library to be aware of the cache directories.
+2. **Globally:** As some library locations may not be writable, we need to have
+   a global out-of-line cache that is used if the first one is not writable.
+   This is located under `$ENSO_DATA` (whose location can be obtained from the
+   `RuntimeDistributionManager`), and is located under the path
+   `$ENSO_DATA/cache/ir/$hash/enso-$version/`, where `$hash` is the `SHA3-224`
+   hash of the tuple `(namespace, library_name, version)`, where
+   `version = SemVer | "local"`.
+
+In each location, the IR is stored with the following assumptions:
+
+- The IR file is located in a directory modelled after its module path, followed
+  by a file named after the module itself with the extension `.ir` (e.g. the IR
+  for `Standard.Base.Data.Vector.enso` is stored in
+  `Standard/Base/Data/Vector.ir`).
+- The [metadata](#metadata-format) file is located in a directory modelled after
+  its module path, followed by a file named after the module itself with the
+  extension `.meta` (e.g. the metadata for `Standard.Base.Data.Vector.enso` is
+  stored in `Standard/Base/Data/Vector.meta`).
+
+Storage of the IR only takes place iff the intended location for that IR is
+_empty_.
+
+### Metadata Format
+
+The metadata is used for integrity checking of the cached IR to prevent loading
+corrupted or out of date data from the cache. It is _exceedingly_ important to
+make the cache decoding forwards-compatible. It should be able to ignore unknown
+fields when loading, allowing us to update the logic in the future if needed.
+
+It is a JSON file as follows:
+
+```typescript
+{
+  sourceHash: String; // The hash of the corresponding source file.
+  blobHash: String; // The hash of the blob.
+  compilationStage: String; // The compilation stage of the IR.
+}
+```
+
+All hashes are encoded in SHA3-224 format, as is used by other components in the
+Engine. The engine version is encoded in the cache path, and hence does not need
+to be explicitly specified in the metadata.
+
+### Portability Guarantees
+
+As part of this design we provide only the following portability guarantees:
+
+- The metadata format must be _extensible_, and be read by _all_ future versions
+  of Enso.
+- The serialised IR must be able to be deserialised by _the same version of
+  Enso_ that wrote the original blob.
+
+## Loading the IR
+
+Loading the IR is a multi-stage process that involves performing integrity
+checking on the loaded cache. It works as follows.
+
+1. **Find the Cache:** Look in the library's `.cache` folder first, checking if
+   there is IR for the module that is valid for the current configuration. If
+   there is no cache here, instead check in the global cache directory under
+   `$ENSO_DATA`. This should be hooked into in `Compiler::parseModule`.
```

Comment:
Also, I'm not sure if the local cache should be prioritized over the global one.

Let's consider a scenario where the library is read-only and its local cache contains an entry, but for some, probably unlikely, reason, this cache entry cannot be used (let's say it is outdated or broken, maybe it was shipped corrupted). Then we will evict this cache, but since it is read-only, we can only write the new cache to global. But if local is preferred, on the next run we will again look at it and fail again.

Unless this process actually looks to local, and if the cache is _present but not 'correct'_ it then checks global too, then this order would make sense too, but it would still be inefficient as it would always first load this local entry, fail with it and only later fallback to global. Whereas with global taking precedence, it will be updated with correct cache after first run and should from now on always hit immediately.